ladybug/datacollection.py

```diff
@@ -455,304 +446,528 @@ def interpolate_data(self, timestep, cumulative=None):
-
-           epw = EPW("c:/ladybug/weatherdata.epw")
-           DBT = epw.dry_bulb_temperature
-           # filter data for when dry bulb temperature is more then 25
-           filtered_DBT = DBT.filter_by_conditional_statement('x > 25')
-           # get the list of time stamps that meet the conditional statement
-           print(filtered_DBT.time_stamps)
-        """
-        def check_input_statement(statement):
-            st_statement = statement.lower() \
-                .replace("and", "").replace("or", "") \
-                .replace("not", "").replace("in", "").replace("is", "")
-
-            parsed_st = [s for s in st_statement if s.isalpha()]
-            if list(set(parsed_st)) != ['x']:
-                statement_error_msg = 'Invalid input statement. ' + \
-                    'Statement should be a valid Python statement' + \
-                    ' and the variable should be named as x'
-                raise ValueError(statement_error_msg)
-
-        check_input_statement(statement)
-
-        statement = statement.replace('x', 'd.value')
-        _filtered_data = [d for d in self.data if eval(statement)]
-
-        # create a new filtered_data
-        if self.header:
-            _filteredHeader = self.header.duplicate()
-            _filteredHeader.analysis_period = None
-            return DataCollection(_filtered_data, _filteredHeader)
+    def _xxrange(self, start, end, step_count):
+        """Generate n values between start and end."""
+        _step = (end - start) / float(step_count)
+        return (start + (i * _step) for i in xrange(int(step_count)))
+
+    def _get_analysis_period_subset(self, a_per):
+        """Return an analysis_period is always a subset of the Data Collection"""
+        if self.header.analysis_period.is_annual:
+            return a_per
+
+        new_needed = False
+        n_ap = [a_per.st_month, a_per.st_day, a_per.st_hour,
+                a_per.end_month, a_per.end_day, a_per.end_hour,
+                a_per.timestep, a_per.is_leap_year]
+        if a_per.st_hour < self.header.analysis_period.st_hour:
+            n_ap[2] = self.header.analysis_period.st_hour
+            new_needed = True
+        if a_per.end_hour > self.header.analysis_period.end_hour:
+            n_ap[5] = self.header.analysis_period.end_hour
+            new_needed = True
+        if a_per.st_time.doy < self.header.analysis_period.st_time.doy:
+            n_ap[0] = self.header.analysis_period.st_month
+            n_ap[1] = self.header.analysis_period.st_day
+            new_needed = True
+        if a_per.end_time.doy > self.header.analysis_period.end_time.doy:
+            n_ap[3] = self.header.analysis_period.end_month
+            n_ap[4] = self.header.analysis_period.end_day
+            new_needed = True
+        if new_needed is False:
+            return a_per
         else:
-            return DataCollection(_filtered_data)
+            return AnalysisPeriod(n_ap[0], n_ap[1], n_ap[2], n_ap[3],
```

Comment:
`AnalysisPeriod(*n_ap)`?

Reply:
Wow, that's a lot cleaner.  Thanks for letting me know about this.